File: Problem-1-Subsets.py
# APPROACH 1:  RECURSIVE
# Time Complexity : O(n * 2^n), n: len(nums)
# Space Complexity : O(n * 2^n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. At each index, two options, maintain current path locally (each time pass a copy of the path)
# 2. If choose -> add path to the global var result, cant choose the same element again. If don't choose -> inc. index. Recursively call the function
# 3. return global var result.

class Solution:
    def __init__(self):
        self.result = []
        
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        if nums is None:
            return None
        
        self.helper(nums, 0, [])
        return self.result
    
    
    def helper(self, nums, index, curr_path):
        if index == len(nums):
            self.result.append(curr_path)
            return
        
        # logic
        # don't choose
        self.helper(nums, index + 1, copy.deepcopy(curr_path))
        
        # choose
        curr_path.append(nums[index])
        self.helper(nums, index + 1, copy.deepcopy(curr_path))
        
      
      
      
      
# APPROACH 2:  BACKTRACK
# Time Complexity : O(n * 2^n), n: len(nums)
# Space Complexity : O(n * 2^n), as we are creating a copy of the current path to add to result
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. Explore all paths for each ind of candidates. At each node, append the current path to the result (not just on reaching the leaf node)
# 2. 1st action -> append current element to path, 2nd -> recurse on the next index (from which pivot again and explore all paths), 3rd -> backtrack (undo the action, remove 
#     the element from the list.
# 3. return the global var result.

class Solution:
    def __init__(self):
        self.result = []
        
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        if nums is None:
            return None
        
        self.helper(nums, 0, [])
        return self.result
    
    
    def helper(self, nums, index, curr_path):
        
        # logic
        self.result.append(copy.deepcopy(curr_path))
        for pos in range(index, len(nums)):
            # action
            curr_path.append(nums[pos])
            
            # recurse
            self.helper(nums, pos + 1, curr_path)
            
            # backtrack
            curr_path.pop()

File: Problem1_Subsets.java
// Time Complexity : o(exponential)
// Space Complexity : o(h) recursive stack space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> subsets(int[] nums) {
        
        result=new ArrayList<List<Integer>>();
        
        helper(nums,0,new ArrayList<Integer>());
        
        
        return result;
    }
    
    private void helper(int[] nums,int index,List<Integer> temp)
    {
        /*
        // one method is to to create new list of temp and pass to avoid deep list or pass existing list but after recursion every call decrement size-1 and store new list only to result.
        if(index==nums.length)
        {
            result.add(temp);
            return;
        }
        
        helper(nums,index+1,new ArrayList<>(temp));
        temp.add(nums[index]);
        helper(nums,index+1,new ArrayList<>(temp));
        
        */
        // another approach using recursion using temp list with out creating new one
        /*
        if(index==nums.length)
        {
            result.add(new ArrayList<>(temp));
            return;
        }
        // dont choose
        helper(nums,index+1,temp);
        // choose
        temp.add(nums[index]);
        helper(nums,index+1,temp);
        // remove last element
        temp.remove(temp.size()-1);
        */
        
        //Aproach using backtracking
        
        result.add(new ArrayList<>(temp));
        for(int i=index;i<nums.length;i++)
        {
            temp.add(nums[i]);
            helper(nums,i+1,temp);
            temp.remove(temp.size()-1);
        }
    }
}